### PR TITLE
Story Hierarchy improved UI

### PR DESCRIPTION
**Pull request by shilman**
Issue: #151

## What I did

Continuation of #1329, but in the main storybook repo

## How to test

Run `cra-kitchen-sink` (see #1329)

### CODE REVIEW

File: lib/ui/src/modules/api/index.js [hunk 1]
```diff
@@ -8,6 +8,7 @@ export default {
       name: 'STORYBOOK',
       url: 'https://github.com/storybooks/storybook',
       sortStoriesByKind: false,
+      resolveStoryHierarchy: storyName => [storyName],
```
Comment (UsulPro) [inline lib/ui/src/modules/api/index.js]: I guess we'll lose a serialized function after passing through the channel. 
We can consider options: 
- pass only JSON data here
- provide this API on the `manager` side (it'll be available in `addons.js`)

Comment (igor-dv) [inline lib/ui/src/modules/api/index.js]: I think I'll just change it to be symbol/regex.
```hierarchySeparator: '/'``` or ```hierarchySeparator: '.'```

**Comment by igor-dv** [comment]
Ok. (Beside the thing I need to fix with what @UsulPro pointed on)

Here is the experimental use of [react-treebeard](https://github.com/alexcurtis/react-treebeard)  (because of [this](https://github.com/storybooks/storybook/pull/1329#issuecomment-310602768))

storybook_hierarchy_react-treebeard

There are warnings because of [this](https://github.com/alexcurtis/react-treebeard/issues/72), but it looks very promising.

File: examples/cra-kitchen-sink/src/stories/index.js [hunk 2]
```diff
@@ -177,16 +177,16 @@ storiesOf('component.Button')
 
 // Atomic
 
-storiesOf('Atoms.Molecules.Cells.simple', module)
+storiesOf('Atoms¯\\_(ツ)_/¯Molecules.Cells/simple', module)
```
Comment (UsulPro) [inline examples/cra-kitchen-sink/src/stories/index.js]: awesome ❤️

**Comment by igor-dv** [comment]
Some updates for the new version
 
react-with-css-with-styleguide-2

I've added Ionicons from the [react-icons](https://github.com/gorangajic/react-icons).

**Comment by UsulPro** [comment]
Looks great! If we got this package in dependencies I can switch to it in #981 as well

**Comment by igor-dv** [comment]
I thought it would be great to have it integrated to theming api too

File: lib/cli/generators/REACT_NATIVE_SCRIPTS/index.js [hunk 3]
```diff
@@ -15,8 +15,7 @@ module.exports = latestVersion('@storybook/react-native').then(version => {
   packageJson.devDependencies['@storybook/react-native'] = `^${version}`;
 
   if (!packageJson.dependencies['react-dom'] && !packageJson.devDependencies['react-dom']) {
-    const reactVersion = packageJson.dependencies.react;
-    packageJson.devDependencies['react-dom'] = reactVersion;
+    packageJson.devDependencies['react-dom'] = '^15.5.4';
```
Comment (shilman) [inline lib/cli/generators/REACT_NATIVE_SCRIPTS/index.js]: hmm?

File: lib/cli/generators/REACT_NATIVE/index.js [hunk 4]
```diff
@@ -31,8 +31,7 @@ module.exports = latestVersion('@storybook/react-native').then(version => {
   packageJson.devDependencies['@storybook/react-native'] = `^${version}`;
 
   if (!packageJson.dependencies['react-dom'] && !packageJson.devDependencies['react-dom']) {
-    const reactVersion = packageJson.dependencies.react;
-    packageJson.devDependencies['react-dom'] = reactVersion;
+    packageJson.devDependencies['react-dom'] = '^15.5.4';
```
Comment (shilman) [inline lib/cli/generators/REACT_NATIVE/index.js]: hmm?

File: examples/cra-kitchen-sink/src/stories/index.js [hunk 5]
```diff
@@ -153,3 +153,43 @@ storiesOf('WithEvents', module)
...
+  .add('first', () => <table><tr><td>first table</td></tr></table>)
+  .add('second', () => <table><tr><td>first table</td></tr></table>);
+
+storiesOf('component.Button')
+  .add('first', () => <button>first button</button>)
+  .add('second', () => <button>first second</button>);
+
+// Atomic
+
+storiesOf('Atoms¯\\_(ツ)_/¯Molecules.Cells/simple', module)
```
Comment (shilman) [inline examples/cra-kitchen-sink/src/stories/index.js]: should be `cells/molecules/atoms`? current order doesn't make sense to me?

File: package.json [hunk 6]
```diff
@@ -58,7 +58,7 @@
     "jest": "^20.0.4",
     "jest-enzyme": "^3.2.0",
     "lerna": "2.0.0-rc.5",
-    "lint-staged": "^4.0.0",
+    "lint-staged": "^3.5.1",
```
Comment (shilman) [inline package.json]: downgrade?

File: lib/cli/generators/REACT_NATIVE_SCRIPTS/index.js (hunk 3 above)
Comment (igor-dv) [inline lib/cli/generators/REACT_NATIVE_SCRIPTS/index.js]: idk what is this ¯\_(ツ)_/¯   ? Maybe it was somehow merged from the previous PR ?

File: examples/cra-kitchen-sink/src/stories/index.js (hunk 5 above)
Comment (igor-dv) [inline examples/cra-kitchen-sink/src/stories/index.js]: I'll fix it.

**Comment by igor-dv** [comment]
Yeah, it's green now =)

**Comment by UsulPro** [comment]
Is it ready to check out?

File: lib/ui/src/modules/ui/components/left_panel/index.js [hunk 7]
```diff
@@ -15,7 +15,13 @@ const mainStyle = {
   padding: '10px 0 10px 10px',
 };
 
-const storyProps = ['stories', 'selectedKind', 'selectedStory', 'onSelectStory'];
+const storyProps = [
+  'storiesHierarchy',
+  'selectedKind',
+  'selectedHierarchy',
+  'selectedStory',
```
Comment (UsulPro) [inline lib/ui/src/modules/ui/components/left_panel/index.js]: Don't we need to have `selectedStory` and `selectedKind` in `propTypes`?

Comment (igor-dv) [inline lib/ui/src/modules/ui/components/left_panel/index.js]: I think linter cries that they are not in use in here. these props just passed further to stories.js where they are actually defined in prop types.

**Comment by igor-dv** [comment]
I think it's ready, but non of the reviewers acepted the PR ...

**Comment by shilman** [comment]
@igor-dv @UsulPro LGTM! Github is not letting me approve the PR because I was the one who opened it 😖

**Comment by shilman** [comment]
@igor-dv I guess we will wait to merge the treebeard improvements until after the new version is ready on NPM?

**Comment by igor-dv** [comment]
@shilman , yes, I think so. I'll merge this one and then will open a new PR for the improvements.

**Comment by joscha** [comment]
references #1445 
references https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...joscha:joscha/fix-storybook-addon-options-hierarchySeparator?expand=1

**Pull request merged**